### PR TITLE
BUG: .rolling() returns incorrect values when ts index is not nano seconds

### LINKED ISSUE #55026: BUG: Rolling pd.date_range incorrect for `unit='s'`
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd

WINDOW = pd.Timedelta(days=4)

def rolling_max(unit: str) -> pd.Series:
    start_date = "2023-01-01"
    end_date = "2023-01-10"

    dates = pd.date_range(start_date, end_date, unit=unit)
    series = pd.Series(0, index=dates)
    series.iloc[0] = 1

    return series.rolling(WINDOW).max()

pd.DataFrame(
    {
        f'{unit=}': rolling_max(unit)
        for unit in ('ns', 's')
    }
).plot(marker='s', grid=True);
```

### Issue Description

Hi,

We've noticed that a `pd.date_range` with `unit='s'` gives incorrect results when we apply a `rolling` function. It appears to return the `expanding` result rather than `rolling`.

### Expected Behavior

The expected result is the same as for `unit='ns'` in the above snippet, I.e. a max of 1 for the first 4 days and 0 beyond that.

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit              : ba1cccd19da778f0c3a7d6a885685da16a072870
python              : 3.11.5.final.0
python-bits         : 64
OS                  : Windows
OS-release          : 10
Version             : 10.0.19044
machine             : AMD64
processor           : Intel64 Family 6 Model 94 Stepping 3, GenuineIntel
byteorder           : little
LC_ALL              : None
LANG                : None
LOCALE              : English_United Kingdom.1252

pandas              : 2.1.0
numpy               : 1.25.2
pytz                : 2023.3.post1
dateutil            : 2.8.2
setuptools          : 68.1.2
pip                 : 23.2.1
Cython              : None
pytest              : None
hypothesis          : None
sphinx              : None
blosc               : None
feather             : None
xlsxwriter          : None
lxml.etree          : None
html5lib            : None
pymysql             : None
psycopg2            : None
jinja2              : None
IPython             : 8.15.0
pandas_datareader   : None
bs4                 : None
bottleneck          : None
dataframe-api-compat: None
fastparquet         : None
fsspec              : None
gcsfs               : None
matplotlib          : 3.7.2
numba               : None
numexpr             : None
odfpy               : None
openpyxl            : None
pandas_gbq          : None
pyarrow             : None
pyreadstat          : None
pyxlsb              : None
s3fs                : None
scipy               : None
sqlalchemy          : None
tables              : None
tabulate            : None
xarray              : None
xlrd                : None
zstandard           : None
tzdata              : 2023.3
qtpy                : None
pyqt5               : None

</details>

### LINKED ISSUE #55106: BUG: .rolling() returns incorrect values when ts index is not nano seconds
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd

df_time = pd.DataFrame({'B': [0, 1, 2, 4, 5, 6]},
                       index=[pd.Timestamp('20130101 09:00:00'),
                              pd.Timestamp('20130101 09:00:02'),
                              pd.Timestamp('20130101 09:00:03'),
                              pd.Timestamp('20130101 09:00:06'),
                              pd.Timestamp('20130101 09:00:07'),
                              pd.Timestamp('20130101 09:00:08')])
sum1 = df_time.rolling('1s').sum()
# micro seconds / milliseconds breaks the correct rolling!
df_time.index = df_time.index.as_unit('us')
sum2 = df_time.rolling('1s').sum()

pd.testing.assert_frame_equal(sum1, sum2)
```

### Issue Description

The rolling function returns incorrect values in case of the index isn't nano seconds.

### Expected Behavior

Both sums should be identical

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 37ea63d540fd27274cad6585082c91b1283f963d
python           : 3.10.9.final.0
python-bits      : 64
OS               : Linux
OS-release       : 6.2.0-31-generic
Version          : #31~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Wed Aug 16 13:45:26 UTC 2
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_IL
LOCALE           : en_IL.UTF-8
pandas           : 2.0.1
numpy            : 1.24.3
pytz             : 2023.3
dateutil         : 2.8.2
setuptools       : 57.5.0
pip              : 22.3.1
Cython           : 3.0.0
pytest           : 7.3.1
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : 3.1.2
lxml.etree       : 4.9.3
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.1.2
IPython          : 8.14.0
pandas_datareader: None
bs4              : 4.12.2
bottleneck       : 1.3.7
brotli           : 1.0.9
fastparquet      : 2023.8.0
fsspec           : 2023.6.0
gcsfs            : 2023.6.0
matplotlib       : 3.7.2
numba            : 0.57.1
numexpr          : 2.8.4
odfpy            : None
openpyxl         : 3.1.2
pandas_gbq       : 0.19.2
pyarrow          : 12.0.1
pyreadstat       : None
pyxlsb           : 1.0.10
s3fs             : None
scipy            : 1.10.1
snappy           : None
sqlalchemy       : None
tables           : 3.8.0
tabulate         : None
xarray           : 2023.7.0
xlrd             : 2.0.1
zstandard        : None
tzdata           : 2023.3
qtpy             : None
pyqt5            : None

</details>

### PR DESCRIPTION
**Pull request by hkhojasteh**
- [X] closes #55106
- [X] closes #55026
- [X] [Tests added and passed]

The bug is referred to and fixed inside the code. It needs to be generalised. @jorisvandenbossche we need to have the first feedback if we can generalise the data type with some utils or more generally. Done during the in-person sprint check.

### CODE REVIEW

File: doc/source/conf.py [hunk 1]
```diff
@@ -172,7 +172,9 @@
 import pandas  # isort:skip
 
 # version = '%s r%s' % (pandas.__version__, svn_version())
-version = str(pandas.__version__)
+# version = str(pandas.__version__)
+
+version = str(pandas.__version__).split("+")[0]
```
Comment (jorisvandenbossche) [inline doc/source/conf.py:177]: Can you revert this change here? (that's from the other PR, because of making this branch after you worked on the other PR from the main branch)

File: pandas/core/window/rolling.py [hunk 2]
```diff
@@ -1884,6 +1884,9 @@ def _validate(self):
             else:
                 self._win_freq_i8 = freq.nanos
 
+            if self._on.dtype == "M8[us]":
+                self._win_freq_i8 /= 1000
+
```
Comment (jorisvandenbossche) [inline pandas/core/window/rolling.py:1889]: @jbrockmendel to generalize this, we could easily add two more `elif dtype == "M8[ms/s]"` checks and do this division explicitly for each of the three non-nanoseconds resolutions, but do you know of a better pattern? Is there some helper function that could be used here?

Or otherwise another way to convert the `freq` to the correct resolution instead of calling `freq.nanos`?

File: pandas/tests/window/test_rolling.py [hunk 3]
```diff
@@ -1950,3 +1950,24 @@ def test_numeric_only_corr_cov_series(kernel, use_arg, numeric_only, dtype):
...
+        {"B": [0, 1, 2, 4, 5, 6]},
+        index=[
+            Timestamp("20130101 09:00:00"),
+            Timestamp("20130101 09:00:02"),
+            Timestamp("20130101 09:00:03"),
+            Timestamp("20130101 09:00:06"),
+            Timestamp("20130101 09:00:07"),
+            Timestamp("20130101 09:00:08"),
+        ],
+    )
```
Comment (jorisvandenbossche) [inline pandas/tests/window/test_rolling.py:1967]: I know this is the example from the issue report, but now we are writing the test, we could actually slightly simplify the creation here, to simplify the test code:

```suggestion
    df = DataFrame({"A": range(5)}, index=date_range("2013-01-01", freq="1s", periods=5))
```

File: pandas/tests/window/test_rolling.py [hunk 4]
```diff
@@ -1950,3 +1950,24 @@ def test_numeric_only_corr_cov_series(kernel, use_arg, numeric_only, dtype):
         op2 = getattr(rolling2, kernel)
         expected = op2(*arg2, numeric_only=numeric_only)
         tm.assert_series_equal(result, expected)
+
+
+def test_rolling_rolling_sum_window_microseconds_confilict_timestamp():
```
Comment (jorisvandenbossche) [inline pandas/tests/window/test_rolling.py:1955]: When generalizing the fix for all resolutions, we can update this test to be parametrized across all resolutions:

```

```suggestion
@pytest.mark.parametrize("unit", ["s", "ms", "us", "ns"])
def test_rolling_rolling_sum_window_microseconds_confilict_timestamp(unit):
```

and then this `unit` variable can be used in `as_unit` or when constructing the object

**Comment by hkhojasteh** [comment]
Great! I’ll fix it soon :)

File: pandas/tests/window/test_rolling.py (hunk 4 above)
Comment (hkhojasteh) [reply pandas/tests/window/test_rolling.py:1955]: Great! I’ll fix it soon :)

File: pandas/core/window/rolling.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/window/rolling.py:1889]: maybe `Timedelta(freq).as_unit(correct_unit)._value`

**Comment by hkhojasteh** [comment]
The test was parametrised, the rolling unit generalised and the version commit reverted :)
@jorisvandenbossche please let me know in case of any other changes

**Comment by hkhojasteh** [comment]
@lithomas1 kindly this fix is also related to other subfunctions like max. More at https://github.com/pandas-dev/pandas/issues/55026. The PR with the related tests are available.

**Comment by lithomas1** [comment]
Can you add a whatsnew note for 2.1.2?

**Comment by hkhojasteh** [comment]
> Can you add a whatsnew note for 2.1.2?

Gladly added :)

File: doc/source/whatsnew/v2.1.2.rst [hunk 5]
```diff
@@ -15,6 +15,7 @@ Fixed regressions
 ~~~~~~~~~~~~~~~~~
 - Fixed bug in :meth:`DataFrame.resample` where bin edges were not correct for :class:`~pandas.tseries.offsets.MonthBegin` (:issue:`55271`)
 - Fixed bug where PDEP-6 warning about setting an item of an incompatible dtype was being shown when creating a new conditional column (:issue:`55025`)
+- Fixed bug in :class:`pandas.core.window.Rolling` where df.rolling does not work with 'datetime64[us]', 'datetime64[ms]', and 'datetime64[s]' index types for rolling functions (:issue:`55026`, :issue:`55106`, :issue:`55299`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.1.2.rst:18]:
```suggestion
- Fixed bug in :class:`pandas.core.window.Rolling` where non-nanosecond ``on`` would produce incorrect results (:issue:`55026`, :issue:`55106`, :issue:`55299`)
```

File: pandas/tests/window/test_rolling.py [hunk 6]
```diff
@@ -1950,3 +1950,38 @@ def test_numeric_only_corr_cov_series(kernel, use_arg, numeric_only, dtype):
         op2 = getattr(rolling2, kernel)
         expected = op2(*arg2, numeric_only=numeric_only)
         tm.assert_series_equal(result, expected)
+
+
+@pytest.mark.parametrize("unit", ["s", "ms", "us", "ns"])
+def test_rolling_rolling_sum_window_microseconds_confilict_timestamp(unit):
+    # GH#55106
+    df_time = DataFrame(
+        {"A": range(5)}, index=date_range("2013-01-01", freq="1s", periods=5)
```
Comment (mroeschke) [inline pandas/tests/window/test_rolling.py:1959]: Could you also test where `date_range` has `tz="UTC"` using `pytest.mark.parameterize`?

File: pandas/tests/window/test_rolling.py [hunk 7]
```diff
@@ -1950,3 +1950,38 @@ def test_numeric_only_corr_cov_series(kernel, use_arg, numeric_only, dtype):
         op2 = getattr(rolling2, kernel)
         expected = op2(*arg2, numeric_only=numeric_only)
         tm.assert_series_equal(result, expected)
+
+
+@pytest.mark.parametrize("unit", ["s", "ms", "us", "ns"])
+def test_rolling_rolling_sum_window_microseconds_confilict_timestamp(unit):
```
Comment (mroeschke) [inline pandas/tests/window/test_rolling.py:1956]:
```suggestion
def test_rolling_rolling_sum_window_microseconds_conflict_timestamp(unit):
```

File: pandas/tests/window/test_rolling.py [hunk 8]
```diff
@@ -1950,3 +1950,38 @@ def test_numeric_only_corr_cov_series(kernel, use_arg, numeric_only, dtype):
...
+    )
+    sum_in_nanosecs = df_time.rolling("1s").sum()
+    # micro seconds / milliseconds should not breaks the correct rolling
+    df_time.index = df_time.index.as_unit(unit)
+    sum_in_microsecs = df_time.rolling("1s").sum()
+    sum_in_microsecs.index = sum_in_microsecs.index.as_unit("ns")
+    tm.assert_frame_equal(sum_in_nanosecs, sum_in_microsecs)
+
+@pytest.mark.parametrize("unit", ["s", "ms", "us", "ns"])
+def test_rolling_rolling_max_window_nanoseconds_confilict_timestamp(unit):
```
Comment (mroeschke) [inline pandas/tests/window/test_rolling.py:1969]:
```suggestion
def test_rolling_rolling_max_window_nanoseconds_conflict_timestamp(unit):
```

File: pandas/tests/window/test_rolling.py (hunk 6 above)
Comment (hkhojasteh) [reply pandas/tests/window/test_rolling.py:1959]: Surprisingly seems it has another issue by `tz="UTC"`. Should I make another issue and continue there? Meanwhile am investigating to see how it is happening

Comment (mroeschke) [reply pandas/tests/window/test_rolling.py:1959]: If we can fix this within this PR that would be good

Comment (hkhojasteh) [reply pandas/tests/window/test_rolling.py]: I have added the patch considering the `UTC` for `_index_array` since it is returning the wrong values after casting `idx` for non-nanosecond values :) Have not found any other way at the moment to consider these values. I'm open to any other suggestions for handling the inner values in a more organised way.

Comment (jorisvandenbossche) [reply pandas/tests/window/test_rolling.py]: What you added now feels like a stop-gap solution, converting the non-nanoseconds datetimes to nanoseconds (which will only work as long as they are not out of bounds), while we should try to fix it properly by not having the convert the actual data, just as you did for tz-naive data. But I think we can fix this at the same place where you are construction the correct `win_freq_i8`, will comment there.

File: pandas/core/window/rolling.py [hunk 9]
```diff
@@ -405,6 +405,15 @@ def _index_array(self):
...
             idx = cast("PeriodIndex | DatetimeIndex | TimedeltaIndex", self._on)
+            if (type(idx) == DatetimeIndex) and (idx.T.tz is not None):
+                if idx.T.unit == "s":
+                    return idx.asi8 * int(1e9)
+                elif idx.T.unit == "ms":
+                    return idx.asi8 * int(1e6)
+                elif idx.T.unit == "us":
+                    return idx.asi8 * int(1e3)
+                elif idx.T.unit == "ns":
+                    return idx.asi8
```
Comment (jorisvandenbossche) [inline pandas/core/window/rolling.py:416]: With my other comment below, this can be removed again, I think

File: pandas/core/window/rolling.py [hunk 10]
```diff
@@ -1889,6 +1898,13 @@ def _validate(self):
             else:
                 self._win_freq_i8 = freq.nanos
 
+            if self._on.dtype == "M8[us]":
+                self._win_freq_i8 /= 1e3
+            elif self._on.dtype == "M8[ms]":
+                self._win_freq_i8 /= 1e6
+            elif self._on.dtype == "M8[s]":
+                self._win_freq_i8 /= 1e9
```
Comment (jorisvandenbossche) [inline pandas/core/window/rolling.py:1906]: So the problem here is that we are checking with hardcoded "M8[s]" types, and so this `if` check fails if you have tz-aware data:

```
(Pdb) self._on.dtype
datetime64[s, UTC]
(Pdb) self._on.dtype == "M8[us]"
False
```

So we should generalize this check. I think we can do something like

```suggestion
            else:
                self._win_freq_i8 = freq.nanos
                unit = dtype_to_unit(self._on.dtype)
                if unit == "us":
                    self._win_freq_i8 /= 1e3
                elif unit == "ms":
                    self._win_freq_i8 /= 1e6
                elif unit == "s":
                    self._win_freq_i8 /= 1e9
```

Because we know here that we have a `datetime64` dtype here (numpy or pd.DatetimeTzDtype), this should work. This `dtype_to_unit` function is currently defined in `pandas.core.arrays.datetimelike` (we should maybe move that, but that can be done later, for now fine to import it from there)

File: pandas/tests/window/test_rolling.py [hunk 11]
```diff
@@ -1950,3 +1950,41 @@ def test_numeric_only_corr_cov_series(kernel, use_arg, numeric_only, dtype):
         op2 = getattr(rolling2, kernel)
         expected = op2(*arg2, numeric_only=numeric_only)
         tm.assert_series_equal(result, expected)
+
+
+@pytest.mark.parametrize("unit", ["s", "ms", "us", "ns"])
+@pytest.mark.parametrize("tz", [None, "UTC"])
+def test_rolling_rolling_sum_window_microseconds_conflict_timestamp(unit, tz):
```
Comment (jorisvandenbossche) [inline pandas/tests/window/test_rolling.py:1957]:
```suggestion
def test_rolling_timedelta_window_non_nanoseconds(unit, tz):
```

Let's simplify the test name a bit here (it's not specific about microseconds or sum; but the test is now generalized to test all resolutions, and it is specific about using a timedelta window (not fixed number of rows))

File: pandas/tests/window/test_rolling.py [hunk 12]
```diff
@@ -1950,3 +1950,41 @@ def test_numeric_only_corr_cov_series(kernel, use_arg, numeric_only, dtype):
...
+    # microseconds / milliseconds should not break the correct rolling
+    df_time.index = df_time.index.as_unit(unit)
+    sum_in_microsecs = df_time.rolling("1s").sum()
+    sum_in_microsecs.index = sum_in_microsecs.index.as_unit("ns")
+    tm.assert_frame_equal(sum_in_nanosecs, sum_in_microsecs)
+
+
+@pytest.mark.parametrize("unit", ["s", "ms", "us", "ns"])
+@pytest.mark.parametrize("tz", [None, "UTC"])
+def test_rolling_rolling_max_window_nanoseconds_conflict_timestamp(unit, tz):
```
[the remark this answers is not in the document]
Comment (hkhojasteh) [reply pandas/tests/window/test_rolling.py:1972]: Yes, exactly
I have added that due to the request here in the thread related to #55026 which I investigated. I can add the `max` to the current test for this PR then

**Comment by hkhojasteh** [comment]
@jorisvandenbossche I have modified the validation function based on your suggestion. Tests are merged and the previous conversation has been removed. Please check it out and let me know

File: pandas/core/window/rolling.py [hunk 13]
```diff
@@ -1889,6 +1891,14 @@ def _validate(self):
             else:
                 self._win_freq_i8 = freq.nanos
 
+                unit = dtype_to_unit(self._on.dtype)
+                if unit == "us":
+                    self._win_freq_i8 /= 1e3
```
Comment (jbrockmendel) [inline pandas/core/window/rolling.py:1896]: should this use floordiv?

Comment (jbrockmendel) [reply pandas/core/window/rolling.py:1896]: you might be better off doing this higher up instead of doing `freq.nanos` above (which could raise for very large freq anyway) doing `Timedelta(freq).as_unit(unit)._value`

Comment (hkhojasteh) [reply pandas/core/window/rolling.py:1896]: Thanks for the review and the suggestion. Didn't get it quite well in which situation could make an issue? Could I have the suggested code with the direct modification in a comment, please?

Comment (jorisvandenbossche) [reply pandas/core/window/rolling.py:1896]: IIUC, it would be like `self._win_freq_i8 = Timedelta(freq).as_unit(unit)._value`, and then you don't need the if/else checks.

Comment (hkhojasteh) [reply pandas/core/window/rolling.py]: Great! That's a util function we were originally searching for :) I have updated the code

**Comment by hkhojasteh** [comment]
@jorisvandenbossche I was about to ask how is the merge on this fix going on? Anything that I can help with?

**Comment by lithomas1** [comment]
Tests are failing

```
FAILED pandas/tests/window/test_rolling.py::tests_empty_df_rolling[1s] - TypeError: cannot get datetime metadata from non-datetime type
FAILED pandas/tests/window/test_timeseries_window.py::TestRollingTS::test_invalid_window_nonfixed[2MS] - AssertionError: Regex pattern did not match.
 Regex: '\\<2 \\* MonthBegins\\> is a non-fixed frequency'
 Input: 'Value must be Timedelta, string, integer, float, timedelta or convertible, not MonthBegin'
FAILED pandas/tests/window/test_timeseries_window.py::TestRollingTS::test_invalid_window_nonfixed[freq1] - AssertionError: Regex pattern did not match.
 Regex: '\\<2 \\* MonthBegins\\> is a non-fixed frequency'
 Input: 'Value must be Timedelta, string, integer, float, timedelta or convertible, not MonthBegin'
FAILED pandas/tests/window/test_timeseries_window.py::TestRollingTS::test_rolling_on_empty - TypeError: cannot get datetime metadata from non-datetime type
```

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @hkhojasteh and @jorisvandenbossche
